compiler/ast.nim

```diff
@@ -869,6 +872,10 @@ type
       bitsize*: int
       alignment*: int # for alignment
     else: nil
+    lazyDecl*: PSym
```

Comment:
why is this required?

Reply:
it turns out, surprisingly, that so far the hardest thing to address was the case of existing fwd declarations, and making sure everything keeps working when those are present (note that they're not needed after this PR, but the code should still work when those are defined in existing code)

the difficulty is:

```
proc fn(a: int) = discard # distractor
proc fn(a: Foo, b = bar()) # fwd decl
proc fn(a: Foo, b = bar()) = discard # impl decl
```
lazy semcheck registers those 3 overloads without any semcheck (in particular bar doesn't need semchecking yet), and I don't know any way to know that 2 and 3 represent same symbol until semcheck happens, which happens once some code refers to `fn`, at which point the semcheck for each overload (in the same scope only) is done

back to your question:

I'm using `lazyDecl` to resolve an apparent openSymChoice into just a single Psym in cases like this, see regression test:
```nim
    block:
      proc fnAux(): int
      proc fn8(r = fnAux) = discard
      proc fnAux(): int = discard
      fn8()
```
i detect that `n[1].lazyDecl = n[0]` (1 is a fwd declaration of the other) without having to re-compute that the overloads collapse.

I could use a side-channel to store this relationship though, in finalized PR.

Reply:
That's what I suspected, but why not just ignore forward decls? These are the ones without body that are *not* .magic or .importc/cpp/js

Reply:
i don't see how these can be ignored, for several reasons; one of those reasons is that fwd decls can contain more information than the implementation:

```nim
  proc baz(): int = 12
  proc fn1*(a: int, b = baz()) {.gcsafe.}
  proc fn1(a: int, b: int) = echo (a,b)
```

if you ignore the fwd decl, you'd miss:
* gcsafe
* the export
* the optional param

Reply:
Hmm, good points.